src/error_measures.jl

```diff
@@ -0,0 +1,165 @@
+
+    # both of these are computed in the initialize! routine
+    initial_mean::Float64 = 0.0
+    initial_vorticity::Float64 = 0.0
+
+    timestep_counter::Int = 0
+    time::Vector{DateTime} = []
+    l1error::Vector{Float64} = []
+    l2error::Vector{Float64} = []
+    l∞error::Vector{Float64} = []
+    mean::Vector{Float64} = []
+    variance::Vector{Float64} = []
+    minimum::Vector{Float64} = []
+    maximum::Vector{Float64} = []
+end
+
+"""
+$(TYPEDSIGNATURES)
+Error measures for the test cases of Williamson et al. 1992, J Computational Physics"""
+function SpeedyWeather.initialize!(callback::ErrorMeasures,
+                                   progn::PrognosticVariables,
+                                   diagn::DiagnosticVariables,
+                                   model::AbstractModel)
+
+    n = progn.clock.n_timesteps + 1
+    callback.time = zeros(DateTime, n)
+    callback.l1error = zeros(n)
+    callback.l2error = zeros(n)
+    callback.l∞error = zeros(n)
+    callback.mean = zeros(n)
+    callback.variance = zeros(n)
+    callback.minimum = zeros(n)
+    callback.maximum = zeros(n)
+
+    η0 = diagn.grid.pres_grid
+    H = model.atmosphere.layer_thickness
+    Hb = model.orography.orography
+    h0 = @. η + H - Hb
+
+    callback.initial_mean = ∬dA(h0)
+    callback.initial_variance = ∬dA((h0 .- callback.initial_mean) .^ 2)
+
+    callback.time[1] = progn.clock.time
+    callback.timestep_counter = 1
+
+    l1, l2, l∞, mean, var, min, max = error_measures(callback, diagn, model)
+
+    callback.l1error[1] = l1
+    callback.l2error[1] = l2
+    callback.l∞error[1] = l∞
+    callback.mean[1] = mean
+    callback.variance[1] = var
+    callback.minimum[1] = min
+    callback.maximum[1] = max
+
+end
+
+"""
+$(TYPEDSIGNATURES)
+Error measures for the test cases of Williamson et al. 1992, J Computational Physics"""
+function SpeedyWeather.callback!(callback::ErrorMeasures,
+                                 progn::PrognosticVariables,
+                                 diagn::DiagnosticVariables,
+                                 model::AbstractModel)
+
+    callback.timestep_counter += 1
+
+    callback.time[i] = progn.clock.time
+
+    l1, l2, l∞, mean, var, min, max = error_measures(callback, diagn, model)
+
+    callback.l1error[i] = l1
+    callback.l2error[i] = l2
+    callback.l∞error[i] = l∞
+    callback.mean[i] = mean
+    callback.variance[i] = var
+    callback.minimum[i] = min
+    callback.maximum[i] = max
+end
+
+using NCDatasets
+
+"""
+$(TYPEDSIGNATURES)
+Error measures for the test cases of Williamson et al. 1992, J Computational Physics"""
+function SpeedyWeather.finalize!(callback::ErrorMeasures,
+                                 progn::PrognosticVariables,
+                                 diagn::DiagnosticVariables,
+                                 model::AbstractModel)
+
+    n_timesteps = callback.timestep_counter
+
+    ds = NCDataset(joinpath(pwd(), callback.output_file), "c")
```

Comment:
you can create the netcdf at `finalize!` but you can also write into it directly instead of writing into the vectors in the `ErrorMeasures` callback and then use those to write the netCDF file. Both works!